### PR TITLE
Record training outcomes in the train event

### PR DESCRIPTION
**Pull request by glenn-jocher**
Follow-on to #25440, which did this for inference. Training fired at `on_train_start` with task, model and device — we learned that a training *started*, never what it produced. This moves the event to `on_train_end` and carries the result.

## The event

`fitness` (best, task-agnostic — mAP50-95 for detect), `epochs_done`, `hours`, `n`, `data`, `imgsz`, `batch`, `optimizer`, `lr0`, `arch`, `GPU`, and nine toggles packed into one `flags` param (`pretrained,cos_lr,amp,rect,multi_scale,freeze,dropout,early_stop,resume,ddp`). 24 params max on CUDA, under GA4's hard 25 limit. Measured 23 on CPU.

Values are recorded **as applied, not as requested**, wherever the two differ — otherwise unrelated runs merge into one cell:

- `optimizer` — `auto` (the default) picks AdamW or MuSGD by iteration count
- `lr0` — read from `optimizer.param_groups[0]["initial_lr"]`. `auto` fits its own lr0 and logs that it is *ignoring* `cfg.lr0`; on a default run `cfg.lr0` says `0.01` while the optimizer runs at `0.000119`
- `batch` — autobatch and OOM retries both move it
- `amp` — `check_amp()` turns a requested `True` off on unsupported hardware
- `arch` — via `unwrap_model()`, since DDP and `torch.compile` both wrap away the `.yaml`

`n` is the train split size, deliberately the same param name predict uses for images seen, so one column means "images this run saw" across modes. `data` is `Path.stem`, so `coco8.yaml` and a `coco8/` directory land in one cell.

## Drive-by fix: training has been sending a duplicate event

`final_eval()` runs a standalone validator built with `copy(trainer.args)`, so `validator.args.mode` is still `"train"` while `validator.training` is `False` — exactly what the old `on_val_start` guard tested. Since an event is named for its mode, that sent a **second event named `train`**, indistinguishable from the first. Measured on main, a 36 s training:

```
FIRE task=detect|train_start  flushed=True
FIRE task=detect|val_start    flushed=True
GA4 receives: train, train      <- two events, one training run
```

Runs under the 30 s rate limit send one, runs over it send two, so existing training counts are inflated toward 2x for real runs in a way that tracks duration and can't be corrected after the fact. It also would have silently killed this PR: the phantom fires first, claims the rate-limit slot, and the new event is enqueued and lost at exit.

The guard is now `validator.args.mode == "val"`, which is the predicate the original was reaching for — `Model.val()` sets `mode` explicitly, so user val and `benchmark()` still fire.

## Verification

Every claim here was measured against a local collector standing in for GA4, on real runs.

- **Delivery 8/8.** The daemon POST survives `on_train_end` — `_clear_memory()`, `loader.close()`, `unset_deterministic()` and teardown follow it, so training does not pay the cost that made predict's hook marginal.
- **One event per run**, on both a 5 s and a 36 s training (the latter sent two on main).
- **Payload confirmed** across detect, classify, early-stop, `val=False`, flags-heavy, DDP-wrapped, resumed and dict-`data` runs.
- **No regression**: predict 22 params, val 12, export 13, all unchanged. `_arch()` still resolves on the AutoBackend path.
- `tests/test_engine.py` 33 passed; `test_python.py` train/val/predict/export subset 45 passed.

Findings from an adversarial review pass, all confirmed and fixed before this PR: the `lr0` error above; `arch` returning `None` for every multi-GPU run; `data` emitting a path fragment for grounding datasets that pass a dict (`{'path': '/home/bobsmith', ...}` yielded `"bobsmith', 'train': 'a', ...}"` — a username, verbatim, so `data` is dropped unless it is a str/Path); and resumed runs pairing an absolute epoch count with session-only wall clock.

## Two things worth knowing

**Crashes are now invisible.** Firing at the end means an OOM or divergence sends nothing, where main sent at start. Defensible — a crashed run has no accuracy — but it skews the corpus toward runs that finished, and OOM rate is itself worth knowing. Deliberate, and easy to revisit.

**Sweep delivery is not guaranteed by a long-lived process.** The queue only flushes on a *subsequent* `events()` call; there is no flush at exit. A process's last queued event is lost unless its call came at least 30 s after the previous flush. Three fast trials in one process delivered 1/3; two realistic ~36 s trials delivered 2/2. Real tuner trials are fine, sub-30 s trials silently are not. Not addressed here — an atexit flush is new machinery and outside this PR.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
📈 Improved analytics event tracking for training, validation, prediction, and model architecture reporting.

### 📊 Key Changes
- Moved training event submission from `on_train_start` to `on_train_end`, ensuring final metrics, duration, fitness, and resolved settings are available.
- Added detailed training analytics, including:
  - Dataset, image size, completed epochs, batch size, training duration, and dataset size.
  - Fitness, optimizer, learning rate, architecture, GPU, and training flags such as AMP, pretrained weights, resume, DDP, and early stopping.
- Updated event handling to pass a completed `run` object for both trainers and predictors.
- Improved architecture detection for trainers, predictors, wrapped models, and exported models.
- Prevented duplicate training events during a trainer’s final validation by only sending validation events for explicit validation runs.
- Updated callback registration and removed the obsolete `on_train_start` documentation entry.
- Continued enforcing analytics safeguards, including exception handling and the 25-parameter event limit.

### 🎯 Purpose & Impact
- ✅ Produces more accurate and useful anonymous analytics based on completed training runs rather than initial settings.
- 📊 Enables better comparison of training performance, configurations, hardware, and model architectures.
- 🚫 Eliminates duplicate or misleading training events generated during final validation.
- 🛡️ Keeps analytics failures isolated so they do not interrupt training, validation, or prediction workflows.
- 🔄 Maintains compatibility with wrapped models such as DDP and EMA while improving architecture identification.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your training outcome event updates, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover the training event changes and NDJSON dataset example, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated review message, and an Ultralytics engineer will assist with any questions or follow-up items. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean PR. The callback move is correctly wired for training, preserves Platform upload behavior, avoids duplicate train events from trainer-owned validation, and reads finalized trainer state after final evaluation. The event fields are guarded so analytics collection cannot interrupt training.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you and congratulations to @glenn-jocher on merging this impactful analytics improvement!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

By capturing completed training runs with final metrics, resolved settings, hardware details, and accurate architecture information—while preventing duplicate events and isolating analytics failures—this PR makes Ultralytics analytics more reliable and actionable without disrupting user workflows. Your thoughtful work helps the community better understand and improve YOLO training performance. 🙌
